mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,828 @@
+        let mut header: MetaDataHeader =
+            deserialize_from(&mut Cursor::new(slice))
+                .context(DeserializeError {})?;
+
+        if header.header_size != MetaDataHeader::METADATA_HEADER_SIZE {
+            return Err(MetaDataError::HeaderSize {
+                size: header.header_size,
+            });
+        }
+
+        if header.signature != [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61]
+        {
+            return Err(MetaDataError::HeaderSignature {});
+        }
+
+        let checksum = header.self_checksum;
+
+        if header.checksum() != checksum {
+            return Err(MetaDataError::HeaderChecksum {});
+        }
+
+        Ok(header)
+    }
+
+    /// Checksum the header with the checksum field itself set to 0
+    pub fn checksum(&mut self) -> u32 {
+        self.self_checksum = 0;
+        self.self_checksum = crc32::checksum_ieee(&serialize(self).unwrap());
+        self.self_checksum
+    }
+
+    /// Generate a new MetaDataHeader based on the partition information and the
+    /// underlying block_size
+    pub fn new(block_size: u32, partition: &GptEntry) -> MetaDataHeader {
+        let index_start = Aligned::get_blocks(
+            MetaDataHeader::METADATA_HEADER_SIZE,
+            block_size,
+        );
+
+        let data_start = index_start
+            + Aligned::get_blocks(
+                MetaDataHeader::MAX_INDEX_ENTRIES
+                    * MetaDataHeader::INDEX_ENTRY_SIZE,
+                block_size,
+            );
+
+        MetaDataHeader {
+            signature: [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61],
+            header_size: MetaDataHeader::METADATA_HEADER_SIZE,
+            self_checksum: 0,
+            generation: 0,
+            self_lba: partition.ent_start + 1, /* skip the first block of the
+                                                * partition */
+            index_start: index_start as u64,
+            used_entries: 0,
+            max_entries: MetaDataHeader::MAX_INDEX_ENTRIES,
+            entry_size: MetaDataHeader::INDEX_ENTRY_SIZE,
+            index_checksum: 0,
+            data_start: data_start as u64,
+            data_end: partition.ent_end - partition.ent_start - 1,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Clone)]
+pub struct MetaDataIndexEntry {
+    /// Current object revision
+    pub revision: u64,
+    /// When this entry was created (number of microseconds since UNIX_EPOCH)
+    pub timestamp: u128,
+    /// CRC-32 checksum of all the blocks comprising the stored object
+    pub data_checksum: u32,
+    /// Location of first data block (relative to self_lba in the
+    /// MetaDataHeader)
+    pub data_start: u64,
+    /// Location of last data block (relative to self_lba in the
+    /// MetaDataHeader)
+    pub data_end: u64,
+}
+
+impl MetaDataIndexEntry {
+    /// Convert a slice into an index array
+    pub fn from_slice(
```

Comment:
tip, these days rust has this: https://doc.rust-lang.org/std/convert/trait.TryFrom.html

The difference here is that the commonly known Trait from<T> for U -- can not fail, as it simply returns U. TryFrom, returns a Result<U, E> which means that it can try but is allowed to fail. 

so you *COULD* impl TryFrom for this but in the near future somewhere :)